rfcs/20200616-keras-multihead-attention.md

```diff
@@ -0,0 +1,340 @@
+# RFC: Multihead Attention and EinsumDense on Keras
+
+| Status        | (Proposed / Accepted / Implemented / Obsolete)          |
+| :------------ | :------------------------------------------------------ |
+| **RFC #**     | [260](https://github.com/tensorflow/community/pull/260) |
+| **Author(s)** | Hongkun Yu (hongkuny@google.com)                        |
+| **Sponsor**   | Francois Chollet (fchollet@google.com)                  |
+| **Updated**   | 2020-06-16                                              |
+
+## Objective
+
+Introduce the MultiHeadAttention layer and EinsumDense layer to tf.keras.
+
+## Motivation
+
+MultiHeadAttention is very popular and has become standard for deep learning
+libraries. We propose to contribute a flexible well-defined implementation
+inside Keras absorbing common best practices from reference libraries.
+
+## User Benefit
+
+We can standardize the implementation of Transformer layers and use the best
+practice. We offer a rich set of functionalities to different use cases, e.g.
+different project spaces, outputing multi-head attention scores for analysis,
+etc. We also modularize computations to make the MultiHeadAttention layer
+extensible to variants.
+
+## Design Proposal
+
+### Key Features
+
+*   Returns multi-headed attention scores, which is commonly useful for
+    attention visualization and analysis.
+*   Supports query (Q), key (K), value (V) tensors as individual inputs and
+    supports projecting Q, K, V to different dimensions.
+*   Final outputs projects to user specified dimensions.
+*   Using tf.einsum to express high-dimensional computation and adopts
+    [tf.keras.layers.experimental.EinsumDense](https://www.tensorflow.org/api_docs/python/tf/keras/layers/experimental/EinsumDense)
```

Comment:
What's the relationship between `tfa.layers` and `tf.keras.layers.experimental`? cc @seanpmorgan @karmel @bhack

Reply:
In particular, relation to the existing [tfa.layers.MultiHeadAttention](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/multihead_attention.py).

Reply:
IIUC layers that go in `tf.keras.layers.experimental` are slated to land in keras, but the API is not set in stone.

Addons would be a place for contributions whos broad applicability is not yet clear, or it is mostly used by a smaller subset of the community (Per charter). This gets tricky though because MultiHeadAttention has proven its applicability but there was no installable implementation in the TF ecosystem. Perhaps we should bring all situations like this up to the Keras team beforehand? That adds another subjective step though so a roadmap would be preferable.

Reply:
The addons implementation is a very good reference. We incorporate the features inside the  addons version and generalize more to fit the emerging needs. For this design, it should cover the tfa.layers.MultiHeadAttention. We hope this common layer can be inside tf.keras directly. The implementation started in model garden back to last year.

Reply:
`.experimental` was defined and approved in https://github.com/tensorflow/community/blob/master/governance/api-reviews.md#experimental-apis.
Sometimes when an experimental namespace doesn't get too much traction it could has an opportunity to be [downstreamed to tf.addons](https://github.com/tensorflow/community/issues/242) as an alternative to jubbe removed (but this downstreaming step is not defined in the RFC).

Reply:
For the MultiHeadAttention layer, it is relatively clear that we should put it inside core keras finally.